### PR TITLE
[SPARK-4987] [SQL] parquet timestamp type support

### PR DESCRIPTION
**Pull request by adrian-wang**

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 1]
```diff
@@ -141,6 +142,12 @@ private[sql] trait SQLConf {
     getConf(PARQUET_BINARY_AS_STRING, "false").toBoolean
 
   /**
+   * When set to true, we always treat INT96Values in Parquet files as timestamp.
+   */
+  private[spark] def isParquetINT96AsTimestamp: Boolean =
+    getConf(PARQUET_INT96_AS_TIMESTAMP, "false").toBoolean
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: We don't really use INT96 for anything else (and I don't think other systems do either?) so maybe this should be true by default?

**Comment by marmbrus** [comment]
Thanks for doing this, I've been getting a ton of requests for this feature!

Can you also add this to the sql programming guide?

**Comment by adrian-wang** [comment]
Oh sorry, I just checked Impala's configuration and I think it is not what it is here. I'll change my code to conform to that.

File: sql/core/pom.xml [hunk 2]
```diff
@@ -69,6 +69,11 @@
       <version>2.3.0</version>
     </dependency>
     <dependency>
+      <groupId>org.jodd</groupId>
+      <artifactId>jodd-core</artifactId>
+      <version>${jodd.version}</version>
+    </dependency>
```
Comment (marmbrus) [inline sql/core/pom.xml]: I'm pretty hesitant to add a dependency here as they are very high cost for a project as big as Spark.  Is there any way to do this without adding this?

Comment (adrian-wang) [inline sql/core/pom.xml]: We can also convert to/from Julian by ourselves... I'll draft it,

File: sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala [hunk 3]
```diff
@@ -138,7 +139,13 @@ private[sql] trait SQLConf {
    * When set to true, we always treat byte arrays in Parquet files as strings.
    */
   private[spark] def isParquetBinaryAsString: Boolean =
-    getConf(PARQUET_BINARY_AS_STRING, "false").toBoolean
+    getConf(PARQUET_BINARY_AS_STRING, "true").toBoolean
```
Comment (scwf) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: why this change?

Comment (adrian-wang) [inline sql/core/src/main/scala/org/apache/spark/sql/SQLConf.scala]: Oh, this is a mistake... I'll modify it in next commit. Actually what I meant to change is the following value.
Thanks!

File: sql/core/pom.xml (hunk 2 above)
Comment (adrian-wang) [inline sql/core/pom.xml]: Writing it by hand may be dangerous, because of leap seconds [https://en.wikipedia.org/wiki/Leap_second], as the specific date could be inaccurate.
And by check the pom[http://repo1.maven.org/maven2/org/jodd/jodd-core/3.5.2/jodd-core-3.5.2.pom] of jodd-core, there's no additional dependence, so the influence is comparatively small. Use jodd to covert also make everything consistent with hive, so we can be 100% compatible with those data generated by hive. So I'd prefer keep this.

**Comment by adrian-wang** [comment]
The [NanoTime in Hive-0.14](https://github.com/apache/hive/blob/f64d707e4d59aafbcf7e9b82ed199f6781d97946/ql/src/java/org/apache/hadoop/hive/ql/io/parquet/timestamp/NanoTime.java) is a little bit different from NanoTime in parquet-examples, and here are some [related discussions](https://issues.apache.org/jira/browse/HIVE-7263?focusedCommentId=14038502&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14038502). So I just rewrite `NanoTime` in hive into scala, instead of using NanoTime from parquet-examples.

File: sql/core/pom.xml (hunk 2 above)
Comment (marmbrus) [inline sql/core/pom.xml]: Okay, you are right that its a bad idea to do this by hand.  Are there any dependencies that Spark SQL already has that could be used instead?

Comment (marmbrus) [inline sql/core/pom.xml]: Okay, I talked to @pwendell and we think it would be better to use Joda time if possible since spark already depends on that library in other subprojects.  What do you think?

Comment (adrian-wang) [inline sql/core/pom.xml]: I have read the documents of joda-time. The toJulianDayNumber API only valid since 2.2, but what we use in spark is 2.1.

**Comment by yhuai** [comment]
Can you add unit tests? I tried it and got 
```
java.lang.RuntimeException: unable to convert datatype TimestampType in CatalystConverter
```
I think you need to update `org.apache.spark.sql.parquet.CatalystConverter#createConverter`.

**Comment by adrian-wang** [comment]
Thanks Yin, I have fixed that.

**Comment by felixcheung** [comment]
I've tested this PR but the result seems to be off.
Parquet generated from Hive with timestamp values set by 'from_utc_timestamp('1970-01-01 08:00:00','PST')'

What I see with this PR:
scala> t.take(10).foreach(println(_))
...
15/01/18 22:06:41 INFO NewHadoopRDD: Input split: ParquetInputSplit{part: file:/users/x/parquetwithtimestamp start: 0 end: 25448 length: 25448 hosts: [] requestedSchema: message root {
  optional binary code (UTF8);
  optional binary description (UTF8);
  optional int32 total_emp;
  optional int32 salary;
  optional int96 timestamp;
}
 readSupportMetadata: {org.apache.spark.sql.parquet.row.metadata={"type":"struct","fields":[{"name":"code","type":"string","nullable":true,"metadata":{}},{"name":"description","type":"string","nullable":true,"metadata":{}},{"name":"total_emp","type":"integer","nullable":true,"metadata":{}},{"name":"salary","type":"integer","nullable":true,"metadata":{}},{"name":"timestamp","type":"timestamp","nullable":true,"metadata":{}}]}, org.apache.spark.sql.parquet.row.requested_schema={"type":"struct","fields":[{"name":"code","type":"string","nullable":true,"metadata":{}},{"name":"description","type":"string","nullable":true,"metadata":{}},{"name":"total_emp","type":"integer","nullable":true,"metadata":{}},{"name":"salary","type":"integer","nullable":true,"metadata":{}},{"name":"timestamp","type":"timestamp","nullable":true,"metadata":{}}]}}}
15/01/18 22:06:41 WARN ParquetRecordReader: Can not initialize counter due to context is not a instance of TaskInputOutputContext, but is org.apache.hadoop.mapreduce.task.TaskAttemptContextImpl
15/01/18 22:06:41 INFO InternalParquetRecordReader: RecordReader initialized will read a total of 823 records.
15/01/18 22:06:41 INFO InternalParquetRecordReader: at row 0. reading next block
15/01/18 22:06:41 INFO CodecPool: Got brand-new decompressor [.snappy]
15/01/18 22:06:41 INFO InternalParquetRecordReader: block read in memory in 27 ms. row count = 823
[00-0000,All Occupations,134354250,40690,1974-01-07 17:58:00.000008896]
[11-0000,Management occupations,6003930,96150,1974-01-07 17:58:00.000008896]

Expect: 1970-01-01 08:00:00

Actual: 1974-01-07 17:58:00.000008896

Any idea?

File: sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala [hunk 4]
```diff
@@ -294,14 +294,14 @@ class ParquetQuerySuite extends QueryTest with FunSuiteLike with BeforeAndAfterA
     // Check to make sure that the attributes from either side of the join have unique expression
     // ids.
     query.queryExecution.analyzed.output.filter(_.name == "myint") match {
-      case Seq(i1, i2) if(i1.exprId == i2.exprId) =>
+      case Seq(i1, i2) if i1.exprId == i2.exprId =>
```
Comment (liancheng) [inline sql/core/src/test/scala/org/apache/spark/sql/parquet/ParquetQuerySuite.scala]: Just a remind, `ParquetQuerySuite` is considered deprecated, and I'm going to remove it in #4116. I should have added a comment to make this clear. Didn't remove it at first because several Parquet PRs were out there and might cause unwanted merge conflicts.

File: docs/sql-programming-guide.md [hunk 5]
```diff
@@ -581,6 +581,15 @@ Configuration of Parquet can be done using the `setConf` method on SQLContext or
...
 </tr>
 <tr>
+  <td><code>spark.sql.parquet.int96AsTimestamp</code></td>
+  <td>true</td>
+  <td>
+    Some Parquet-producing systems, in particular Impala, store Timestamp into INT96. Spark would also
+    store Timestamp as INT96 because we need to avoid precision lost of the nanoseconds field. This
+    flag tells Spark SQL to interpret INT96 data as a timestamp to provide compatibility with these systems.
+  </td>
+</tr>
```
Comment (liancheng) [inline docs/sql-programming-guide.md]: I'm probably OK with this property. Just wanna ask doesn't Impala store original type information (TIMESTAMP) together in Parquet metainfo?

Comment (marmbrus) [inline docs/sql-programming-guide.md]: Do they?  As far as I can tell the parquet spec does not have a nanosecond
precision timestamp type.

Comment (liancheng) [inline docs/sql-programming-guide.md]: Oh, I see the difference here. Double checked, Parquet only provides TIMESTAMP and TIMESTAMP_MILLIS

Comment (marmbrus) [inline docs/sql-programming-guide.md]: Yeah, I agree that it's weird though.  Perhaps we should ask the parquet
list why they don't support the int 96 version.

Comment (felixcheung) [inline docs/sql-programming-guide.md]: From my digging only Parquet-format 2.2 has the TIMESTAMP and TIMESTAMP_MILLS types. Cloudera is still on 1.5.0
Hive/Impala has been writing this INT96 nano sec format that's different.

--- Original Message ---

**Comment by adrian-wang** [comment]
@felixcheung  I have tested my code with HIVE 0.14.0 release.
I first tried to generate data with hive but spark need metadata to read hive's data.
Then I generated data using spark, then save as parquet file, copied the saved file to my hdfs to my hive warehouse dir, and in hive cli I executed create table. Then I use select to read the data, it worked fine.

How did you read data from hive? Can you show me how to reproduce this?

**Comment by yhuai** [comment]
@adrian-wang Can you try the parquet file uploaded in https://issues.apache.org/jira/browse/SPARK-4768? Thanks!

**Comment by adrian-wang** [comment]
@yhuai Thanks for the link, I'm afraid the attachment contains only the last row, in which the timestamp field is null. Anyway, I have reproduced the bug @felixcheung mentioned by read my hive generated parquet records. It's strange that Hive can read this PR's output correctly but this PR cannot read Hive's output correctly... I'll look into that.

**Comment by felixcheung** [comment]
Good to hear. Here's how I create my test data, I run this in Hive and then take the data from HDFS directly and Spark is able to read/parse the data file (with issue above):

Set parquet.compression = SNAPPY;

DROP TABLE testdata;

CREATE TABLE testdata
STORED AS PARQUET
AS SELECT a.*, from_utc_timestamp('1970-01-01 08:00:00','PST') as timestamp
FROM sample_07 AS a;


I have looked into this a fair bit and attempted a fix. Thanks for working on this fix and let me know if I could help in any way.

**Comment by adrian-wang** [comment]
I have fixed the bug....
This is quite embarrassing, I forgot to set those factors(NANOS_PER_SECOND, SECONDS_PER_MINUTE, MINUTES_PER_HOUR) to Long when divide, so it overflew... I have tested it, now it works fine.

**Comment by adrian-wang** [comment]
ping

File: pom.xml [hunk 6]
```diff
@@ -149,6 +149,7 @@
     <scala.binary.version>2.10</scala.binary.version>
     <jline.version>${scala.version}</jline.version>
     <jline.groupid>org.scala-lang</jline.groupid>
+    <jodd.version>3.5.2</jodd.version>
```
Comment (pwendell) [inline pom.xml]: Any reason not to use the latest version (3.6.3?)

**Comment by pwendell** [comment]
Dependency looks fine to me, thanks for running it by.

File: pom.xml (hunk 6 above)
Comment (adrian-wang) [inline pom.xml]: To keep align with Hive 0.14.0, not necessary though

**Comment by adrian-wang** [comment]
I just rebased my code and upgrade jodd to 3.6.3.

**Comment by yhuai** [comment]
I tried the newly uploaded parquet data in https://issues.apache.org/jira/browse/SPARK-4768 (I set my timezone to UTC), for one line, I got
```
[test row 5,2015-01-02 20:54:10.000456789]
```
But, the data was generated by 
```
insert into string_timestamp (dummy,timestamp1) values('test row 5', '2015-01-02 20:54:10.123456789');
```
Can you take a look?

**Comment by adrian-wang** [comment]
@yhuai Sorry, I got a misunderstanding on the setNanos API, it works OK now.

**Comment by marmbrus** [comment]
Jenkins, test this please

**Comment by yhuai** [comment]
Jenkins, test this please

**Comment by marmbrus** [comment]
Thanks!  Merged to master.

File: sql/core/pom.xml (hunk 2 above)
Comment (rxin) [inline sql/core/pom.xml]: FYI most systems don't support leap seconds. I'm not sure why we'd want to support them here...

Comment (adrian-wang) [inline sql/core/pom.xml]: This mainly aims to make sure we are compatible with hive

Comment (rxin) [inline sql/core/pom.xml]: Does Hive support leap seconds? I looked into the implementation of jodd -- I don't think it supports that when doing date timestamp conversion. I could be wrong though.

Comment (adrian-wang) [inline sql/core/pom.xml]: If we are going to do this conversion ourselves, I think it is fine...
